mesonbuild/options.py

```diff
@@ -300,78 +319,85 @@ def validate_value(self, value: T.Any) -> bool:
             return False
         raise MesonException(f'Option "{self.name}" value {value} is not boolean (true or false).')
 
-class UserIntegerOption(UserOption[int]):
-    def __init__(self, name: str, description: str, value: T.Any, yielding: bool = DEFAULT_YIELDING,
-                 deprecated: T.Union[bool, str, T.Dict[str, str], T.List[str]] = False):
-        min_value, max_value, default_value = value
-        self.min_value = min_value
-        self.max_value = max_value
-        c: T.List[str] = []
-        if min_value is not None:
-            c.append('>=' + str(min_value))
-        if max_value is not None:
-            c.append('<=' + str(max_value))
-        choices = ', '.join(c)
-        super().__init__(name, description, choices, yielding, deprecated)
-        self.set_value(default_value)
-
-    def validate_value(self, value: T.Any) -> int:
+
+class _UserIntegerBase(UserOption[_T]):
+
+    min_value: T.Optional[int]
+    max_value: T.Optional[int]
+
+    if T.TYPE_CHECKING:
+        def toint(self, v: str) -> int: ...
+
+    def __post_init__(self, value_: _T) -> None:
+        super().__post_init__(value_)
+        choices: T.List[str] = []
+        if self.min_value is not None:
+            choices.append(f'>= {self.min_value!s}')
+        if self.max_value is not None:
+            choices.append(f'<= {self.max_value!s}')
+        self.__choices: str = ', '.join(choices)
+
+    def printable_choices(self) -> T.Optional[T.List[str]]:
+        return [self.__choices]
+
+    def validate_value(self, value: T.Any) -> _T:
         if isinstance(value, str):
-            value = self.toint(value)
+            value = T.cast('_T', self.toint(value))
         if not isinstance(value, int):
             raise MesonException(f'Value {value!r} for option "{self.name}" is not an integer.')
         if self.min_value is not None and value < self.min_value:
             raise MesonException(f'Value {value} for option "{self.name}" is less than minimum value {self.min_value}.')
         if self.max_value is not None and value > self.max_value:
             raise MesonException(f'Value {value} for option "{self.name}" is more than maximum value {self.max_value}.')
-        return value
+        return T.cast('_T', value)
+
+
+@dataclasses.dataclass
+class UserIntegerOption(_UserIntegerBase[int]):
+
+    min_value: T.Optional[int] = None
+    max_value: T.Optional[int] = None
```

Comment:
Given that this class derives from `_UserIntegerBase`, do the dataclass members need to be declared again?

Reply:
I made a choice to not implement them in the `_UserIntegerBase` because the two classes have different default values, which even still, the shared base is a bit of a compromise with code reuse, since technically the Umask option always has a min and max.

Reply:
I see what you mean. Rading the code, when I found `min_value: T.Optional[_T]` in the base class, my brain autocompleted it appending a `= None` as it usually follows members declared `T.Optional`. Thus the redefinition here seemed redundant.